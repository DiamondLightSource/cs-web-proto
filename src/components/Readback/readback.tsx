import React from "react";
import { connectionWrapper } from "../ConnectionWrapper/connectionWrapper";
import { CopyWrapper } from "../CopyWrapper/copyWrapper";
import { AlarmBorder } from "../AlarmBorder/alarmBorder";

import classes from "./readback.module.css";
import { macroWrapper } from "../MacroWrapper/macroWrapper";
import { VType } from "../../vtypes/vtypes";
import { Alarm, alarmOf, AlarmSeverity } from "../../vtypes/alarm";
import { vtypeToString } from "../../vtypes/utils";
import {
  Widget,
  PVWidgetInterface,
  ConnectedWidgetInterface
} from "../Widget/widget";

export interface ReadbackProps {
  connected: boolean;
  value?: VType;
  precision?: number;
  style?: object;
}

function getClass(alarmSeverity: any): string {
  switch (alarmSeverity) {
    case AlarmSeverity.MINOR: {
      return classes.Minor;
    }
    case AlarmSeverity.MAJOR: {
      return classes.Major;
    }
  }
  return classes.Readback;
}

export const Readback = (props: ReadbackProps): JSX.Element => {
  let { connected, value, precision = undefined, style } = props;
  const alarm = alarmOf(value);
  let displayedValue;
  if (!value) {
    displayedValue = "Waiting for value";
  } else {
    displayedValue = vtypeToString(value, precision);
  }
  style = { backgroundColor: "#383838", ...props.style };

  // Change text color depending on connection state
  if (!connected) {
    style = {
      ...style,
      color: "#ffffff"
    };
  }

  return (
    <div
      className={`Readback ${classes.Readback} ${getClass(
        alarm.getSeverity()
      )}`}
      style={style}
    >
      {displayedValue}
    </div>
  );
};

interface ConnectedReadbackProps {
  pvName: string;
  rawPvName?: string;
  precision?: number;
  alarm?: Alarm;
  style?: object;
}

export const ConnectedReadback: React.FC<
  ConnectedReadbackProps
> = connectionWrapper(Readback);

interface ConnectedCopyReadbackProps {
  pvName: string;
  rawPvName?: string;
  precision?: number;
  style?: object;
}

export const CopyReadback = (props: {
  pvName: string;
  rawPvName?: string;
  value: VType;
  connected: boolean;
  precision?: number;
  style?: object;
}): JSX.Element => (
  <CopyWrapper
    pvName={props.pvName}
    rawPvName={props.rawPvName}
    connected={props.connected}
    value={props.value}
    style={props.style}
  >
    <Readback
      connected={props.connected}
      value={props.value}
      precision={props.precision}
      style={props.style}
    ></Readback>
  </CopyWrapper>
);

export const ConnectedCopyReadback: React.FC<
  ConnectedCopyReadbackProps
> = connectionWrapper(CopyReadback);

interface ConnectedStandaloneReadbackProps {
  pvName: string;
  precision?: number;
  style?: object;
}

export const StandaloneReadback = (props: {
  pvName: string;
  rawPvName?: string;
  value: VType;
  connected: boolean;
  precision?: number;
  style?: object;
}): JSX.Element => (
  <CopyWrapper
    pvName={props.pvName}
    rawPvName={props.rawPvName}
    connected={props.connected}
    value={props.value}
  >
    <AlarmBorder connected={props.connected} value={props.value}>
      <Readback
        connected={props.connected}
        value={props.value}
        precision={props.precision}
        style={props.style}
      ></Readback>
    </AlarmBorder>
  </CopyWrapper>
);

export const ConnectedStandaloneReadback: React.FC<
  ConnectedStandaloneReadbackProps
> = macroWrapper(connectionWrapper(StandaloneReadback));

interface ReadbackWidgetProps {
  precision?: number;
}

export const ReadbackWidget = (
  props: ReadbackWidgetProps & PVWidgetInterface
): JSX.Element => {
  return <Widget baseWidget={Readback} {...props} />;
};

export const ConnectedReadbackWidget: React.FC<
  ReadbackWidgetProps & ConnectedWidgetInterface
> = macroWrapper(connectionWrapper(ReadbackWidget));
