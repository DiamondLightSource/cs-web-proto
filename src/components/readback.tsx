import React from "react";
import { useSelector } from "react-redux";
import { useSubscription } from "../hooks/useCs";
import { NType } from "../cs";
import { CsState } from "../redux/store";

export const Readback = (props: {
  pvName: string;
  value: string;
}): JSX.Element => {
  let value = "";
  if (props.value !== null) {
    value = props.value;
  }
  return (
    <div>
      {props.pvName}: {value}
    </div>
  );
};

export const ConnectedReadback = (props: { pvName: string }): JSX.Element => {
  useSubscription(props.pvName);
  const latestValue = useSelector((state: CsState): string => {
    let value = state.valueCache[props.pvName];
    if (value == null) {
      return "";
    } else {
      return value.value.toString();
    }
  });
  return <Readback pvName={props.pvName} value={latestValue} />;
};
