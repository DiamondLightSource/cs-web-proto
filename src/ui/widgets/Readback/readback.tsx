import React from "react";

import { Widget } from "../widget";
import { PVComponent, PVWidgetPropType } from "../widgetProps";

import classes from "./readback.module.css";
import {
  IntPropOpt,
  BoolPropOpt,
  InferWidgetProps,
  ChoicePropOpt,
  FontPropOpt,
  ColorPropOpt,
  BorderPropOpt
} from "../propTypes";
import { registerWidget } from "../register";
import { LabelComponent } from "../Label/label";
import { AlarmQuality, DAlarm } from "../../../types/dtypes";

const ReadbackProps = {
  precision: IntPropOpt,
  showUnits: BoolPropOpt,
  fgAlarmSensitive: BoolPropOpt,
  textAlign: ChoicePropOpt(["left", "center", "right"]),
  transparent: BoolPropOpt,
  font: FontPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  border: BorderPropOpt
};

// Needs to be exported for testing
export type ReadbackComponentProps = InferWidgetProps<typeof ReadbackProps> &
  PVComponent;

function getClass(connected: boolean, alarmSeverity: AlarmQuality): string {
  let cls = classes.Readback;
  if (!connected) {
    cls += ` ${classes.Disconnected}`;
  } else {
    switch (alarmSeverity) {
      case AlarmQuality.WARNING: {
        cls += ` ${classes.Minor}`;
        break;
      }
      case AlarmQuality.ALARM: {
        cls += ` ${classes.Major}`;
        break;
      }
    }
  }
  return cls;
}

export const ReadbackComponent = (
  props: ReadbackComponentProps
): JSX.Element => {
  const {
    connected,
    value,
    precision,
    font,
    foregroundColor,
    backgroundColor,
    border,
    fgAlarmSensitive = false,
    transparent = false,
    textAlign = "center",
    showUnits = false
  } = props;
  // Decide what to display.
  const alarm = value?.getAlarm() || DAlarm.NONE;
  const display = value?.getDisplay();
  let displayedValue;
  if (!value) {
    displayedValue = "Waiting for value";
  } else {
    if (precision !== undefined && !isNaN(value.getDoubleValue())) {
      displayedValue = value.getDoubleValue().toFixed(precision);
    } else {
      displayedValue = value.getStringValue();
    }
  }

  // Add units if there are any and show units is true
  if (showUnits && display?.units) {
    displayedValue = displayedValue + ` ${display.units}`;
  }

  // Handle foreground alarm sensitivity.
  let className = classes.Readback;
  // TODO: should we show disconnection even if not alarm sensitive?
  if (fgAlarmSensitive) {
    className = getClass(connected, alarm.quality);
  }
  // Use a LabelComponent to display it.
  return (
    <LabelComponent
      className={className}
      text={displayedValue}
      transparent={transparent}
      textAlign={textAlign}
      font={font}
      foregroundColor={foregroundColor}
      backgroundColor={backgroundColor}
      border={border}
    ></LabelComponent>
  );
};

const ReadbackWidgetProps = {
  ...ReadbackProps,
  ...PVWidgetPropType
};

export const Readback = (
  props: InferWidgetProps<typeof ReadbackWidgetProps>
): JSX.Element => (
  <Widget
    // TODO: Note that we asking for both string and double here;
    // this subverts the intended efficiency.
    pvType={{ string: true, double: true }}
    baseWidget={ReadbackComponent}
    {...props}
  />
);

registerWidget(Readback, ReadbackWidgetProps, "readback");
