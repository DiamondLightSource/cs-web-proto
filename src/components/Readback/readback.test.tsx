import React from "react";
import { Readback } from "./readback";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

test("snapshot matches", (): void => {
  const readback = renderer.create(<Readback value={"hello"} />);
  let json = readback.toJSON();
  expect(json).toMatchSnapshot();
});

it("renders a basic element", (): void => {
  const wrapper = shallow(<Readback value={"hello"} />);
  expect(wrapper.text()).toEqual("hello");
});

it("applies precision to numbers", (): void => {
  const wrapper = shallow(<Readback value={"3.14159265359"} precision={2} />);
  expect(wrapper.text()).toEqual("3.14");
});
