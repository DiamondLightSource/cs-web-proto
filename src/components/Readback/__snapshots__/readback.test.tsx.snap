// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matches 1`] = `
<div
  className="BaseWidget"
  onClick={[Function]}
  style={Object {}}
>
  <span
    className="tooltiptext"
  >
    mypv
    <br />
    [
    hello, 2019-08-21 11:58:00
    ]
  </span>
  <div>
    mypv
    : 
    hello
  </div>
</div>
`;
